# Poker Probability and Statistics in Python Part 1

Data Scientists create machine learning models to make predictions and optimize decisions. In online poker, the options are whether to bet, call, or fold.  You aren’t allowed to use software to make those decisions though. That's where most online poker sites draw the line in the rules. Since you can't train a machine learning model, you must train your brain. This requires an endless stream of equity calculations away from the poker table.  These equity calculations use many different probability and statistics concepts.  In Part 1 of this tutorial you’ll learn some of these concepts using a deck of cards and generic poker situations.  In Part 2 you’ll apply what you’ve learned to actual poker hands that I played during my career.

For several years, beginning in 2010 I made a living playing online poker professionally. Data Science was a natural progression for me as it requires a similar skill-set as earning a profit from online poker. I wrote a blog about [what data science has in common with poker](https://medium.springboard.com/how-i-used-professional-poker-to-become-a-data-scientist-e49b75dfe8e3), and I mentioned that each time a poker hand is played at an online poker site, a hand history is generated.  These hand histories explain everything that each player did during that hand. I used software called [Hold’em Manager](https://holdemmanager.com/?a_aid=sharpdata) (think Tableau for poker) to take advantage of this data.  Hold'em Manager downloads each of these hand histories in real time to a PostgreSQL database so you can keep track of your opponent’s tendencies.

## Probability Theory

Humans are notoriously bad at intuitively understanding probability theory.  The concepts, however, are very important to for data science and predictive analytics.  Data science is about using data from events that have already occurred to predict future events.  Statistics makes sense the past, while probability theory helps predict the future.  More specifically, probability tells you how often, or the rate at which an event is likely to occur, given multiple trials of the event.  Understanding probability theory isn’t so difficult once you take a closer look.  First, a few things to keep in mind;

- An experiment is a set of rules that governs a specific operation that is being performed;
- A trial is the performance or exercise of that experiment;
- An outcome is the result of a given trial;
- An event is an outcome or any combination of outcomes;
- The probability of an event is always between 0 and 1, with 0 indicating an event has no chance of occurring and 1 indicating an event will always occur.  

To determine the probability of an event occurring, divide the number of possible event outcomes by the sample space.  The sample space is defined as all the possible outcomes. 


$$P(Event) = \frac{Event Outcomes}{Sample Space}$$

There are 52 cards In a standard deck of cards.

### Card Deck Sample Space Image here 

There are 4 Aces in a deck.  If you want to know the probability of drawing an Ace, you divide the number of possible event outcomes (4), by the sample space (52).

$$P(Ace) = \frac{4}{52}$$

Determine the probability of drawing an Ace.

In [1]:
# Ace Probability Code

# Sample Space
cards = 52

# Outcomes
aces = 4

# Divide possible outcomes by the sample set
ace_probability = aces / cards

# Print probability rounded to two decimal places
print(round(ace_probability, 2))

0.08


The probability of drawing an Ace from a standard deck is 0.08.  To determine probability in percentage form, simply multiply by 100.

In [2]:
# Ace Probability Percent Code
ace_probability_percent = ace_probability * 100

# Print probability percent rounded to one decimal place
print(round(ace_probability_percent, 1))

7.7


The probability of drawing an Ace as a percent is 8%.  Now determine the probability of drawing a card that is a;

1. Heart;
2. Face card (Jacks, Queens, and Kings);
3. Queen of Hearts.

First, create a function to find the probability of an event in percentage form.

In [3]:
# Probability Function Code

# Create function that returns probability percent rounded to one decimal place
def event_probability(event_outcomes, sample_space):
    probability = (event_outcomes / sample_space) * 100
    return round(probability, 1)

# Sample Space
cards = 52

# Determine the probability of drawing a heart
hearts = 13
heart_probability = event_probability(hearts, cards)

# Determine the probability of drawing a face card
face_cards = 12
face_card_probability = event_probability(face_cards, cards)

# Determine the probability of drawing the queen of hearts
queen_of_hearts = 1
queen_of_hearts_probability = event_probability(queen_of_hearts, cards)

# Print each probability
print(heart_probability)
print(face_card_probability)
print(queen_of_hearts_probability)

25.0
23.1
1.9


## Combinations & Permutations

Combinations refers to the number of ways a subset of a specified size can be drawn from a given set.  Here is an example with Aces.  There are four Aces in a deck of cards, and these are all the different combinations of pocket Aces;

1. Ace Hearts / Ace Diamonds
2. Ace Hearts / Ace Clubs
3. Ace Hearts / Ace Spades
4. Ace Diamonds / Ace Clubs
5. Ace Diamonds / Ace Spades
6. Ace Clubs / Ace Spades

There are six combinations of pocket Aces.  The equation for determining combinations is:


$$nCk = \frac{nPk}{k!}$$

Where;

- C = Combinations
- n = Size of the set
- k = Size of the subset
- P = Permutations
- ! = Factorial

To find the number of combinations, you first must find the number of permutations.  Permutations are the number of ways a subset of a specified size can be ***arranged*** from a given set.  The equation for determining permutations is:

$$nPk = \frac{n!}{(n - k)!}$$

To find the number of permutations of pocket Aces, you must divide the factorial of the size of the set by the factorial of the size of the set minus the size of the subset.

$$4P2 = \frac{4!}{(4 - 2)!}$$

In [4]:
# Permutations Code
import math
n = 4
k = 2

# Determine permutations and print result
Permutations = math.factorial(n) / math.factorial(k)
print(Permutations)

12.0


To determine the number of combinations, simply divide the number of permutations by the factorial of the size of the subset.  Try finding the number of starting hand combinations that can be dealt in Texas Hold’em.

$$52C2 = \frac{52P2}{2!}$$

In [5]:
# Combinations Code
n = 52
k = 2

# Determine Permutations
Permutations = math.factorial(n) / math.factorial(n - k)

# Determine Combinations and print result
Combinations = Permutations / math.factorial(k)
print(Combinations)

1326.0


## Independent vs. Dependent Events

Independent events are events that don’t impact the probability of the other event/s.  Drawing a card from the deck, replacing the card, shuffling the deck, and then drawing another card is an example of independent events.  The probability of drawing an Ace the first draw is the same as the second.  

Dependent events are events that have an impact on the probability of the other event/s.  Drawing a card from the deck and then drawing a second card from the deck without replacing the first card is an example of a dependent event.  

The probability of drawing an Ace the fist draw is not the same as the probability of drawing an Ace on the second draw.  After the first card is drawn, the sample space has reduced by 1, from 52 to 51.  Depending on what the card was on the first draw, the number of event outcomes may have also changed.  If the card was an Ace, there are now only 3 Aces remaining for the second draw.

You already know the probability of drawing an Ace on the first draw.  Now determine the probability of drawing an Ace on the second draw , if the first card drawn was;

1. A King
2. An Ace

In [6]:
# Dependent Events Code

# Sample Space
cards = 52
cards_drawn = 1 
cards = cards - cards_drawn 

# Determine the probability of drawing an Ace after drawing a King on the first draw
aces = 4
ace_probability1 = event_probability(aces, cards)

# Determine the probability of drawing an Ace after drawing an Ace on the first draw
aces_drawn = 1
aces = aces - aces_drawn
ace_probability2 = event_probability(aces, cards)

# Print each probability
print(ace_probability1)
print(ace_probability2)

7.8
5.9


Here are a few situations common to poker which are relevant to the concept of dependent events.  Before we get started, a little background info is in order.  The game is Texas Hold’em.  Played with a standard 52 card deck, Texas Hold’em is the most popular of all the poker variations.  Each player is dealt two cards to start the hand and will make the best five-card hand possible by using their two cards combined with the five community cards that are dealt throughout the hand.  Cards are dealt in four rounds;

- **Pre-Flop**: Each player is dealt two cards, known as hole cards
- **Flop**: Three community cards are dealt
- **Turn**: One community card is dealt
- **River**: Final community card is dealt

#### Flush Draw

##### Your Hand

#### Images Here

##### Community Cards

#### Images Here

Your on the Turn and you have four cards to an Ace high Flush.  A Flush is a strong poker hand where all five cards are the same suit.  What’s the probability that the last community card, known as the River Card, is a Diamond?

In [7]:
# Flush Draw Code

# Sample Space
cards = 52
hole_cards = 2
turn_community_cards = 4
cards = cards - (hole_cards + turn_community_cards)

# Outcomes
diamonds = 13
diamonds_drawn = 4
# In poker, cards that complete a draw are known as outs
outs = diamonds - diamonds_drawn

#Determine river flush probability
river_flush_probability = event_probability(outs, cards)
print(river_flush_probability)

19.6


There is roughly a 20% chance of hitting your Flush draw on the River.  Here’s another one:

#### Open Ended Straight Draw

##### Your Hand

#### Images Here

##### Community Cards

#### Images Here

Your on the Turn and you have an open ended Straight draw.  A Straight is another strong hand where there are five cards in sequential order.  The Straight draw is open ended because any Eight ( 8, 9, 10, Jack, Queen) or any King (9, 10, Jack, Queen, King) will complete the straight.  What’s the probability that the River card completes the Straight?

In [8]:
# Straight Draw Code

# Sample Space
cards = 52
hole_cards = 2
turn_community_cards = 4
cards = cards - (hole_cards + turn_community_cards)

# Outcomes
eights = 4
kings = 4
outs = eights + kings

# Determine river straight probability
river_straight_probability = event_probability(outs, cards)
print(river_straight_probability)

17.4


There is roughly a 17% chance of hitting your Straight draw on the River.

## Multiple Events

You’ve learned how to calculate probabilities for single events, but what about multiple events?  

#### Mutually Exclusive Events

To calculate the probability (or the union) of any of multiple mutually exclusive events occurring, you simply add together the probability of each event occurring.  

$$P(Event A \cup Event B) = P(Event A) + P(Event B)$$

Drawing a heart or drawing a club are two mutually exclusive events.  A heart is a heart and a club is a club.  To determine the probability of drawing a heart or drawing a club, add the probability of drawing a heart to the probability of drawing a club.

$$P(Heart \cup Club) = (\frac{13}{52}) \times (\frac{13}{52})$$

Determine the probability of the following mutually exclusive events;

1. Drawing a heart or drawing a club;
2. Drawing an ace, a king or a queen.

In [9]:
# Mutual Exclusivity Code

# Sample Space
cards = 52

# Calculate the probability of drawing a heart or a club
hearts = 13
clubs = 13
heart_or_club = event_probability(hearts, cards) + event_probability(clubs, cards)

# Calculate the probability of drawing an ace, king, or a queen
aces = 4
kings = 4
queens = 4
ace_king_or_queen = event_probability(aces, cards) + event_probability(kings, cards) + event_probability(queens, cards)

print(heart_or_club)
print(ace_king_or_queen)

50.0
23.1


#### Non Mutually Exclusive Events

Not all events are mutually exclusive.  Drawing a heart or drawing an ace are two non mutually exclusive events.  The ace of hearts is both an ace and a heart.  When events are not mutually exclusive, you must correct for the overlap.

$$P(Event A \cup Event B) = P(Event A) + P(Event B) - P(EventA \cup EventB)$$

To calculate the probability of drawing a heart or an ace, add the probability of drawing a heart to the probability of drawing an ace and then subtract the probability of drawing the ace of hearts.

$$P(Heart \cup Ace) = (\frac{13}{52}) + (\frac{4}{52}) - (\frac{1}{52})$$

Calculate the probability of the following non mutually exclusive events;

1. Drawing a heart or an ace;
2. Drawing a red card or drawing a face card.

In [10]:
# Non Mutual Exclusivity Code

# Sample Space
cards = 52

# Calculate the probability of drawing a heart or an ace
hearts = 13
aces = 4
ace_of_hearts = 1
heart_or_ace = event_probability(hearts, cards) + event_probability(aces, cards) - event_probability(ace_of_hearts, cards)

# Calculate the probability of drawing a red card or a face card
red_cards = 26
face_cards = 12
red_face_cards = 6
red_or_face_cards = event_probability(red_cards, cards) + event_probability(face_cards, cards) - event_probability(red_face_cards, cards)

print(round(heart_or_ace, 1))
print(round(red_or_face_cards, 1))

30.8
61.6


#### Intersection of Independent Events

The probability of the intersection of two independent events is determined by multiplying the probabilities of each event occurring.

$$P(Event A \cap Event B) = P(Event A) \times P(Event B)$$

If you want to know the probability of drawing an Ace from a deck of cards, replacing it, reshuffling the deck, and drawing another Ace, you multiply the probability of drawing and Ace times the probability of drawing an Ace.

$$P(Ace \cap Ace) = (\frac{4}{52}) \times (\frac{4}{52})$$

In [11]:
# Two Independent Events Code

# Sample Space
cards = 52

# Outcomes
aces = 4

# Probability of one ace
ace_probability = aces / cards

# Probability of two consecutive independant aces 
two_aces_probability = ace_probability * ace_probability

# Two Ace Probability Percent Code
two_ace_probability_percent = two_aces_probability * 100
print(round(two_ace_probability_percent, 1))

0.6


The probability of drawing two Aces in a row, independently, is 0.592%.  What if the second event is dependant?

#### Intersection of Dependent Events

The probability of the intersection of two non independent events (Event A & Event B given A) is determined by multiplying the probability of Event A occurring times the probability of Event B given A. 

$$P(Event A \cap Event B | A) = P(Event A) \times P(Event B | A)$$

The best starting hand you can have in Texas Hold’em is pocket Aces.  What is the probability of being dealt two Aces?

$$P(Ace \cap Ace | Ace) = (\frac{4}{52}) \times (\frac{3}{51})$$


In [12]:
# Intersection of Dependent Events Code

# Sample Space first draw
cards = 52

# Outcomes first draw
aces = 4

# Probability of ace on first draw
first_ace_probability = aces / cards

# Sample Space second draw
cards = cards - 1

# Outcomes second draw
aces = aces - 1

# Probability of ace on second draw after ace on first
second_ace_probability = aces / cards

# Probability of two consecutive aces (dependent)
both_aces_probability = first_ace_probability * second_ace_probability * 100
print(both_aces_probability)

0.4524886877828055


The probability of drawing two non independent Aces in a row is 0.452%.  Let’s take a look at a couple situations where this comes into play at the poker table.

#### Flop Flush Draw

##### Your Hand

#### Images Here

##### Community Cards

#### Images Here

This is a very similar situation as the Flush draw above, except this time you’re on the flop and have two more community cards to come instead of just one.  How can you determine the probability of getting a Flush by the River?  First you need to determine all the different possible scenarios;

- A) Turn Diamond, River Non Diamond (Made Flush)
- B) Turn Non Diamond, River Diamond (Made Flush)
- C) Turn Diamond, River Diamond (Made Flush)
- D) Turn Non Diamond, River Non Diamond (No Flush)

Those are the only four possibilities, and each of those scenarios are mutually exclusive.  This means that if you add the probabilities of each of those scenarios occurring, the total will be 1.  In other words, one of those four scenarios is definitely going to occur.  You want to know the probability of scenario A, B, or C occurring.  The simplest approach to figure this out is to determine the probability of scenario D, and subtract that from 1.

In [13]:
# Flop Ace Flush Draw Code

# Sample Space on turn
cards = 52
hole_cards = 2
flop_community_cards = 3
cards = cards - (hole_cards + flop_community_cards)

# Outcomes
diamonds = 13
diamonds_drawn = 4
non_diamonds_drawn = 1
outs = diamonds - diamonds_drawn
turn_non_diamonds = cards - outs - non_diamonds_drawn 

# Probability of not getting a diamond on the turn
no_diamond_turn_probability = turn_non_diamonds / cards

# Sample Space on river
turn_community_card = 1
cards = cards - turn_community_card

# Outcomes on river
river_non_diamonds = turn_non_diamonds - turn_community_card

# Probability of not getting a diamond on the river
no_diamond_river_probability = river_non_diamonds / cards

# Probability of not getting a flush
no_flush_probability = no_diamond_turn_probability * no_diamond_river_probability

# Probability of getting a flush
flush_probability = 1 - no_flush_probability
flush_probability_percent = flush_probability * 100

# Print probability percent rounded to one decimal place
print(round(flush_probability_percent, 1))

38.4


Now let's change the hand slightly.

##### Your Hand

#### Images Here

##### Community Cards

#### Images Here

This is a very similar situation as the last one, except for one big difference.  You still have a flush draw, but this time you don’t have the Ace.  If a Diamond falls on the turn and the river, there is a good chance someone will have a better flush draw.  Determine the probability of a Diamond falling on the Turn or the River, but not both.  You already have the probability of not hitting a flush.  Now find the probability of a diamond falling on the turn and river, add that to the probability of not hitting a flush, and subtract from 1.

In [14]:
# Flop 45 Flush Draw Code

# Sample Space on turn
cards = 52
hole_cards = 2
flop_community_cards = 3
cards = cards - (hole_cards + flop_community_cards)

# Outcomes on turn
diamonds = 13
diamonds_drawn = 4
outs = diamonds - diamonds_drawn

# Probability of diamond on turn
diamond_turn_probability = outs / cards

# Sample Space on river
turn_diamonds = 1
cards = cards - turn_diamonds

# Outcomes on river
river_diamonds = outs - turn_diamonds

# Probability of diamond on river
diamond_river_probability = river_diamonds / cards

# Probability of getting a diamond on the turn and river
two_diamonds_probability = diamond_turn_probability * diamond_river_probability

# Determine the probability of getting only one diamond by the river
one_diamond_probability = 1 - (no_flush_probability + two_diamonds_probability)
one_diamond_probability_percent = one_diamond_probability * 100

# Print probability percent rounded to one decimal place
print(round(one_diamond_probability_percent, 1))

35.1


## Expected Value

Every decision made in life boils down to a decision between two or more expected values.  If you’re at the grocery store and deciding between the name brand cereal and the generic brand, you’re making a decision between two expected values.  Does the amount of value you gain from buying the name brand justify the higher cost?  In this case the value gained is intrinsic, and hard to calculate.

You can calculate Expected Value at the poker table though. You simply multiply the Total Value times the probability of winning to get your Expected Value.   

$$Expected Value = Total Value \times Probability$$

What is the expected value if there is \$100 (Total Value) in the pot, and your probability of winning the pot is 0.75?

$$Expected Value = \$100 \times 0.75$$

In [15]:
# EV Example Code
pot = 100
probability = 0.75

# Determine expected value
expected_value = pot * probability
print(expected_value)

75.0


You’re expected value is \$75.  Expected value is an important concept in poker.  Let’s go back to the first flush example to see how to use expected values to your advantage.

##### Your Hand

#### Images Here

##### Opponents Hand

#### Images Here

##### Community Cards

#### Images Here

- Total Pot = \$60

- Opponents Bet = \$20

Your opponent has decided to be helpful and show you his cards, and has a set of 2s.  To win the hand on the River, you must hit any Diamond except a Jack or 2.  The Jack or 2 of Diamonds would give your opponent a better hand, a full house and four of a kind respectively.  You have to call \$20 to stay in the hand, and if you win the hand you win \$60.  If your expected value is greater than \$20 you should call the bet, and if not you should fold.  Figure out if you should call the bet.  

In [16]:
# EV Flush Code

# Sample Space
cards = 52
hole_cards = 2
opponents_hole_cards = 2 # Your opponent provided you information... use it
turn_community_cards = 4
cards = cards - (hole_cards + opponents_hole_cards + turn_community_cards)

# Outcomes
diamonds = 13
diamonds_drawn = 4
diamond_non_outs = 2 # You can't count the two diamonds that won't help you win
outs = diamonds - diamonds_drawn - diamond_non_outs

# Determine win probability
win_probability = outs / cards

# Determine expected value
pot = 60
ev = pot * win_probability

# Print ev and appropriate decision
call_amount = 20
if ev >= 20:
    print(round(ev, 2), 'Call')
else:
    print(round(ev, 2), 'Fold')

9.55 Fold


You’re expected value is only \$9.55, which is less than the \$20 you would need to risk to get that reward, so you must fold.

This concludes Part 1 of the tutorial.  You learned about several core probability concepts including Independent/Dependant events, Multiple events, Expected Values, and Permutations/Combinations and how to calculate each of them.  In Part 2 you will apply these concepts to actual poker hands that I played during my career.


\*One card is dealt face down, known as the Burn card, before the Flop, Turn, and River.  Since the card is dealt face down, and no player knows what it is, it does not count as a trial.